libuavcan/src/transport/uc_can_io.cpp

```diff
@@ -1,330 +1,272 @@
-        registerRejectedFrame();
+    CanTxQueueEntry *entry = new(praw) CanTxQueueEntry(frame, tx_deadline, qos, flags);
+    UAVCAN_ASSERT(entry);
+    auto result = AvlTree::insert(entry);
 
-        // Find a frame with lowest QoS
-        Entry* p = queue_.get();
-        if (p == UAVCAN_NULLPTR)
-        {
-            UAVCAN_TRACE("CanTxQueue", "Push rejected: Nothing to replace");
-            return;
-        }
-        Entry* lowestqos = p;
-        while (p)
-        {
-            if (lowestqos->qosHigherThan(*p))
-            {
-                lowestqos = p;
-            }
-            p = p->getNextListNode();
-        }
-        // Note that frame with *equal* QoS will be replaced too.
-        if (lowestqos->qosHigherThan(frame, qos))           // Frame that we want to transmit has lowest QoS
-        {
-            UAVCAN_TRACE("CanTxQueue", "Push rejected: low QoS");
-            return;                                         // What a loser.
-        }
-        UAVCAN_TRACE("CanTxQueue", "Push: Replacing %s", lowestqos->toString().c_str());
-        remove(lowestqos);
-        praw = allocator_.allocate(sizeof(Entry));        // Try again
+    if (!result) {
+        /* AVL Tree could not allocate a new node */
+        UAVCAN_TRACE("CanTxQueue", "Push rejected: OOM (AvlTree::Node)");
+        safeIncrementRejectedFrames();
+        CanTxQueueEntry::destroy(entry, this->allocator_);
     }
+}
 
-    if (praw == UAVCAN_NULLPTR)
-    {
-        return;                                            // Seems that there is no memory at all.
+void CanTxQueue::remove(CanTxQueueEntry* entry){
+    if (entry == UAVCAN_NULLPTR) {
+        return;
     }
-    Entry* entry = new (praw) Entry(frame, tx_deadline, qos, flags);
-    UAVCAN_ASSERT(entry);
-    queue_.insertBefore(entry, PriorityInsertionComparator(frame));
+
+    // Make the AvlTree remove the specific entry deleting it's Node *
+    this->AvlTree::remove(entry);
+    // Then let the entry destroy it's own contents
+    CanTxQueueEntry::destroy(entry, this->allocator_);
 }
 
-CanTxQueue::Entry* CanTxQueue::peek()
-{
-    const MonotonicTime timestamp = sysclock_.getMonotonic();
-    Entry* p = queue_.get();
-    while (p)
-    {
-        if (p->isExpired(timestamp))
-        {
-            UAVCAN_TRACE("CanTxQueue", "Peek: Expired %s", p->toString().c_str());
-            Entry* const next = p->getNextListNode();
-            registerRejectedFrame();
-            remove(p);
-            p = next;
-        }
-        else
-        {
-            return p;
-        }
+CanTxQueueEntry *CanTxQueue::peek(){
+    auto maxNode = searchForNonExpiredMax(root);
+
+    if(maxNode == UAVCAN_NULLPTR){
+        return UAVCAN_NULLPTR;
     }
-    return UAVCAN_NULLPTR;
+
+    return maxNode->data;
 }
 
-void CanTxQueue::remove(Entry*& entry)
-{
-    if (entry == UAVCAN_NULLPTR)
-    {
-        UAVCAN_ASSERT(0);
-        return;
+// TODO: Isn't that NOT (topPriorityHigherOrEqual) ?
```

Comment:
I read this method as `top_priority_frame >= rhs_frame` so I think this is correct.

Reply:
It's `!rhs_frame.priorityHigherThan(peek_entry->frame)` => `rhs_frame > top_priority_non_expired_frame`. Do I change the name then?

It also returns `false` when the queue is empty.